### PR TITLE
1275 pull integration for permission table

### LINKED ISSUE #1275: Pull integration for permission table
- [ ] Move permissions from `service/src/sync/init_programs_data.rs` to the central server (once server supports it). Inject permissions via script.
- [x] pull permissions table 

There are two sources for permissions:
1) Permissions received during first login of a user
2) Dynamic permissions table rows received through the new pull

Currently, during first user login the user permission table is fully cleared for the user (in `service/src/user_account.rs` `permission_repo.delete_by_user_id(&user.id)?;`)

- [x] make sure to only delete permission entries without `context`, i.e. only entries that come straight from the central server during login.

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #1275 

# 👩🏻‍💻 What does this PR do? 
Sets up user_permission table (which hasn't been created yet in central) sync translation for pull/delete. Filter out any permissions without context on login, so those permissions don't get deleted. Integration tests.

# 🧪 How has/should this change been tested? 
erm....

## 💌 Any notes for the reviewer?
Will need to create script to inject the data into a central datafile.

### CODE REVIEW

File: server/service/src/sync/test/integration/central/user_permission.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            "permission": "DocumentQuery",
+            "context": user_permission_row_2.context,
+        });
+
+        result.push(TestStepData {
+            central_upsert: json!({
+                "user_permission": [user_permission_row_1_json, user_permission_row_2_json],
+            }),
+            central_delete: json!({}),
+            integration_records: IntegrationRecords::new(),
```
Comment (clemens-msupply) [inline server/service/src/sync/test/integration/central/user_permission.rs:54]: The rows are missing here, aren't they?

File: server/service/src/sync/translations/mod.rs [hunk 2]
```diff
@@ -88,10 +90,12 @@ pub(crate) mod LegacyTableName {
     pub(crate) const NAME_STORE_JOIN: &str = "name_store_join";
     pub(crate) const CLINICIAN: &str = "clinician";
     pub(crate) const CLINICIAN_STORE_JOIN: &str = "clinician_store_join";
+    pub(crate) const USER_PERMISSION: &str = "user_permission";
```
Comment (clemens-msupply) [inline server/service/src/sync/translations/mod.rs:93]: this table will most likely be called `om_user_permission`

File: server/repository/src/db_diesel/user_permission_row.rs [hunk 3]
```diff
@@ -105,9 +105,19 @@ impl<'a> UserPermissionRowRepository<'a> {
 
     pub fn delete_by_user_id(&self, user_id: &str) -> Result<(), RepositoryError> {
```
Comment (clemens-msupply) [inline server/repository/src/db_diesel/user_permission_row.rs:106]: Think a better name would be good here, e.g. `delete_basic_permissions_by_user_id()`.

Or even better have a `UserPermissionRepository::delete(filter)` method. Do we do this somewhere already? hopefully straight forward to implement?

File: server/service/src/user_account.rs [hunk 4]
```diff
@@ -67,8 +67,9 @@ impl<'a> UserAccountService<'a> {
                 let user_store_repo = UserStoreJoinRowRepository::new(con);
                 let permission_repo = UserPermissionRowRepository::new(con);
 
-                // remove existing user (if exists)
+                // remove all permissions that don't have context
```
Comment (clemens-msupply) [inline server/service/src/user_account.rs:70]: Having a test that only basic permissions are deleted would be a nice to have!

Comment (roxy-dao) [reply server/service/src/user_account.rs:70]: [3777ff1](https://github.com/openmsupply/open-msupply/pull/1345/commits/3777ff1aa8789b588d1062d2d3a69caa94f0da76)

**Review by andreievg — COMMENTED** [review]
I think we should try to minimise logic in database layer (repository), if we do have helpers like this I think they need to be named to represent the `hidden` functionality (that we are deleting only permissions without context). 

I thought I commented on a line, this is in reference to `delete_by_user_id` on permission_row in repository

**Comment by clemens-msupply** [comment]
> I think we should try to minimise logic in database layer (repository), if we do have helpers like this I think they need to be named to represent the `hidden` functionality (that we are deleting only permissions without context).
> 
> I thought I commented on a line, this is in reference to `delete_by_user_id` on permission_row in repository.
> 
> Also I've noticed test for it is in service but logic in repository ¯_(ツ)_/¯

We actually had the idea to have a repo method to delete by filter (like query) but this wasn't as easy as hoped for and would have required some more refactoring... But yes the name should reflect it a bit better.

File: server/service/src/user_account.rs [hunk 5]
```diff
@@ -67,12 +67,17 @@ impl<'a> UserAccountService<'a> {
...
+                        .has_context(false),
+                )?;
+                for permission in permissions_to_delete {
+                    permission_repo.delete(&permission.id)?;
+                }
                 user_store_repo.delete_by_user_id(&user.id)?;
-                user_repo.delete_by_id(&user.id)?;
-                // insert user
-                user_repo.insert_one(&user)?;
+                user_repo.upsert_one(&user)?;
```
Comment (andreievg) [inline server/service/src/user_account.rs:79]: nicely caught !

File: server/service/src/user_account.rs [hunk 6]
```diff
@@ -67,12 +67,17 @@ impl<'a> UserAccountService<'a> {
                 let user_store_repo = UserStoreJoinRowRepository::new(con);
                 let permission_repo = UserPermissionRowRepository::new(con);
 
-                // remove existing user (if exists)
-                permission_repo.delete_by_user_id(&user.id)?;
+                let permissions_to_delete = UserPermissionRepository::new(con).query_by_filter(
```
Comment (andreievg) [inline server/service/src/user_account.rs:70]: Thanks for the change

**Review by andreievg — COMMENTED** [review]
Would like to approve, but I see integration test may fail (as per your comment on telegram) Due to user not syncing, are you adding the fix for it in this PR ?

**Comment by roxy-dao** [comment]
> Would like to approve, but I see integration test may fail (as per your comment on telegram) Due to user not syncing, are you adding the fix for it in this PR ?

I can do it in this one or in a new PR, I don't mind either way (:

**Review by andreievg — APPROVED** [review]
Thanks, looks good, looks like integration tests will be done in a new issue https://github.com/openmsupply/open-msupply/issues/1393 👍

**Pull request merged**